### PR TITLE
one output_file_grp, make_file_id, assert_file_grp_cardinality

### PR DESCRIPTION
**Pull request by kba**

### CODE REVIEW

File: ocrd_anybaseocr/cli/ocrd_anybaseocr_binarize.py [hunk 1]
```diff
@@ -100,19 +105,17 @@ def dshow(self, image, info):
...
+        assert_file_grp_cardinality(self.output_file_grp, 1)
         oplevel = self.parameter['operation_level']
 
         for (n, input_file) in enumerate(self.input_files):            
+            file_id = input_file.ID.replace(self.input_file_grp, page_grp)
             page_id = input_file.pageId or input_file.ID
             
             LOG.info("INPUT FILE %i / %s", n, page_id)
             pcgts = page_from_file(self.workspace.download_file(input_file))
+            pcgts.set_PcGtsId(file_id)
```
Comment (bertsky) [inline ocrd_anybaseocr/cli/ocrd_anybaseocr_binarize.py]:
```suggestion
            pcgts.set_pcGtsId(file_id)
```

File: ocrd_anybaseocr/cli/ocrd_anybaseocr_block_segmentation.py [hunk 2]
```diff
@@ -139,19 +136,15 @@ def process(self):
...
                 LOG.warning('Operation level %s, but should be "page".', oplevel)
                 break
-            file_id = input_file.ID.replace(self.input_file_grp, page_grp)
 
-            # Use input_file's basename for the new file -
-            # this way the files retain the same basenames:
-            if file_id == input_file.ID:
-                file_id = concat_padded(page_grp, n)
+            file_id = make_file_id(input_file, self.output_file_grp)
+            pcgts.set_PcGtsId(file_id)
```
Comment (bertsky) [inline ocrd_anybaseocr/cli/ocrd_anybaseocr_block_segmentation.py]:
```suggestion
            pcgts.set_pcGtsId(file_id)
```

File: ocrd_anybaseocr/cli/ocrd_anybaseocr_cropping.py [hunk 3]
```diff
@@ -465,20 +462,14 @@ def process(self):
...
                     'Operation level %s, but should be "page".', oplevel)
-            file_id = input_file.ID.replace(
-                self.input_file_grp, page_grp)
-
-            # Use input_file's basename for the new file -
-            # this way the files retain the same basenames:
-            if file_id == input_file.ID:
-                file_id = concat_padded(page_grp, n)
+            file_id = make_file_id(input_file, self.output_file_grp)
+            pcgts.set_PcGtsId(file_id)
```
Comment (bertsky) [inline ocrd_anybaseocr/cli/ocrd_anybaseocr_cropping.py]:
```suggestion
            pcgts.set_pcGtsId(file_id)
```

File: ocrd_anybaseocr/cli/ocrd_anybaseocr_deskew.py [hunk 4]
```diff
@@ -123,19 +126,14 @@ def process(self):
...
                 break
             
-            # Use input_file's basename for the new file -
-            # this way the files retain the same basenames:
-            file_id = input_file.ID.replace(self.input_file_grp, page_grp)
-            if file_id == input_file.ID:
-                file_id = concat_padded(page_grp, n)
-            
+            file_id = make_file_id(input_file, self.output_file_grp)
+            pcgts.set_PcGtsId(file_id)
```
Comment (bertsky) [inline ocrd_anybaseocr/cli/ocrd_anybaseocr_deskew.py]:
```suggestion
            pcgts.set_pcGtsId(file_id)
```

File: ocrd_anybaseocr/cli/ocrd_anybaseocr_dewarp.py [hunk 5]
```diff
@@ -156,18 +164,14 @@ def process(self):
...
                         model, dataset, page, region_xywh, region.id, input_file, orig_img_size, n)
 
-            # Use input_file's basename for the new file -
-            # this way the files retain the same basenames:
-            page_grp = self.output_file_grp
-            file_id = input_file.ID.replace(self.input_file_grp, page_grp)
-            if file_id == input_file.ID:
-                file_id = concat_padded(page_grp, n)
+            file_id = make_file_id(input_file, self.output_file_grp)
+            pcgts.set_PcGtsId(file_id)
```
Comment (bertsky) [inline ocrd_anybaseocr/cli/ocrd_anybaseocr_dewarp.py]:
```suggestion
            pcgts.set_pcGtsId(file_id)
```

File: ocrd_anybaseocr/cli/ocrd_anybaseocr_textline.py [hunk 6]
```diff
@@ -107,18 +102,14 @@ def process(self):
         
                     self._process_segment(region_image, page, region, region_xywh, region.id, input_file, k)
 
-            # Use input_file's basename for the new file -
-            # this way the files retain the same basenames:
-            file_id = input_file.ID.replace(self.input_file_grp, self.output_file_grp)            
-            if file_id == input_file.ID:
-                file_id = concat_padded(self.output_file_grp, n)                
+            file_id = make_file_id(input_file, self.output_file_grp)
+            pcgts.set_PcGtsId(file_id)
```
Comment (bertsky) [inline ocrd_anybaseocr/cli/ocrd_anybaseocr_textline.py]:
```suggestion
            pcgts.set_pcGtsId(file_id)
```

File: ocrd_anybaseocr/cli/ocrd_anybaseocr_tiseg.py [hunk 7]
```diff
@@ -114,18 +111,15 @@ def process(self):
...
                 break
         
-            # Use input_file's basename for the new file -
-            # this way the files retain the same basenames:
-            file_id = input_file.ID.replace(self.input_file_grp, self.output_file_grp)            
-            if file_id == input_file.ID:
-                file_id = concat_padded(self.output_file_grp, n)                
+
+            file_id = make_file_id(input_file, self.output_file_grp)
+            pcgts.set_PcGtsId(file_id)
```
Comment (bertsky) [inline ocrd_anybaseocr/cli/ocrd_anybaseocr_tiseg.py]:
```suggestion
            pcgts.set_pcGtsId(file_id)
```

**Review by bertsky — CHANGES_REQUESTED** [review]
Have not looked at everything yet, but wrong spelling here stands out.

Note: should also change the version to a new major!

File: ocrd_anybaseocr/cli/ocrd_anybaseocr_binarize.py (hunk 1 above)
Comment (kba) [reply ocrd_anybaseocr/cli/ocrd_anybaseocr_binarize.py]: argh, looks like I lost a commit somewhere fixing those. thanks for spotting.

**Comment by kba** [comment]
Except for dewarping (not possible w/o CUDA), every processor should at least run and produce output.

**Comment by bertsky** [comment]
> Except for dewarping (not possible w/o CUDA), every processor should at least run and produce output.

Great work!

For the record, could you identify what specifically is missing for CPU fallback in dewarping?
I remember @VChristlein gave us the following advice:
- https://gitter.im/OCR-D/Lobby?at=5ef6709ffa0c9221fc6192bc
- https://gitter.im/OCR-D/Lobby?at=5ef6713bbb149531edd2661a
- https://gitter.im/OCR-D/Lobby?at=5ef67358c4bdd8347509e8a8
- https://gitter.im/OCR-D/Lobby?at=5ef7131847fdfd21edeaa136

**Comment by kba** [comment]
This is how the `test_dewarp.py` test fails for me w/o the CUDA check skip:

```
/home/kba/build/github.com/OCR-D/monorepo/ocrd_all/ocrd_anybaseocr/tests/test_dewarp.py:32:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
/home/kba/build/github.com/OCR-D/monorepo/ocrd_all/venv/local/sub-venv/headless-tf22/lib/python3.6/site-packages/ocrd/processor/base.py:61: in run_processor
    processor.process()                                                                                                                                                     
/home/kba/build/github.com/OCR-D/monorepo/ocrd_all/ocrd_anybaseocr/ocrd_anybaseocr/cli/ocrd_anybaseocr_dewarp.py:152: in process                                             
    model, dataset, page, page_xywh, page_id, input_file, orig_img_size, n)                                                                                                 
/home/kba/build/github.com/OCR-D/monorepo/ocrd_all/ocrd_anybaseocr/ocrd_anybaseocr/cli/ocrd_anybaseocr_dewarp.py:182: in _process_segment
    data['label'], data['inst'], data['image'])                                                                                                                             
/home/kba/build/github.com/OCR-D/monorepo/ocrd_all/ocrd_anybaseocr/ocrd_anybaseocr/pix2pixhd/models/pix2pixHD_model.py:198: in inference                                     
    input_label, inst_map, real_image, _ = self.encode_input(Variable(label), Variable(inst), image, infer=True)
/home/kba/build/github.com/OCR-D/monorepo/ocrd_all/ocrd_anybaseocr/ocrd_anybaseocr/pix2pixhd/models/pix2pixHD_model.py:113: in encode_input                                 
    input_label = label_map.data.cuda()                                                                                                                                     
/home/kba/build/github.com/OCR-D/monorepo/ocrd_all/venv/local/sub-venv/headless-tf22/lib/python3.6/site-packages/torch/cuda/__init__.py:186: in _lazy_init
    _check_driver()                                                                                                                                                         
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
                                                                                                                                                                             
    def _check_driver():                                                                                                                                                     
        if not hasattr(torch._C, '_cuda_isDriverSufficient'):     
            raise AssertionError("Torch not compiled with CUDA enabled")                           
        if not torch._C._cuda_isDriverSufficient():      
            if torch._C._cuda_getDriverVersion() == 0:                                       
                # found no NVIDIA driver on the system                                                                                                                      
                raise AssertionError("""                                   
    Found no NVIDIA driver on your system. Please check that you                                   
    have an NVIDIA GPU and installed a driver from                
>   http://www.nvidia.com/Download/index.aspx""")                                                                                                                           
E               AssertionError:                                                               
E               Found no NVIDIA driver on your system. Please check that you                                                                                                
E               have an NVIDIA GPU and installed a driver from                                                                                                              
E               http://www.nvidia.com/Download/index.aspx  
```

**Pull request merged**